torchdata/nodes/map.py

```diff
@@ -0,0 +1,132 @@
+# pyre-unsafe
+# import multiprocessing as mp
+import queue
+import threading
+from typing import Callable, Iterator, List, Literal, Optional, TypeVar, Union
+
+import torch.multiprocessing as mp
+
+from torch._utils import ExceptionWrapper
+
+from torchdata.nodes import BaseNode, T
+
+from ._apply_udf import _apply_udf
+
+from ._populate_queue import _populate_queue
+
+
+X = TypeVar("X")
+
+
+class Mapper(BaseNode[T]):
+    def __init__(
+        self,
+        source: BaseNode[X],
+        map_fn: Callable[[X], T],
+    ):
+        self.source = source
+        self.map_fn = map_fn
+
+    def iterator(self) -> Iterator[T]:
+        for item in self.source:
+            yield self.map_fn(item)
+
+
+class ParallelMapper(BaseNode[T]):
+    def __init__(
+        self,
+        source: BaseNode[X],
+        map_fn: Callable[[X], T],
+        num_workers: int,
+        in_order: bool = True,
+        method: Literal["thread", "process"] = "thread",
+    ):
+        self.source = source
+        self.udf = map_fn
+        self.num_workers = num_workers
+        self.in_order = in_order
+
+        self.in_q: Union[queue.Queue, mp.Queue] = queue.Queue() if method == "thread" else mp.Queue()
+        self.out_q: Union[queue.Queue, mp.Queue] = queue.Queue() if method == "thread" else mp.Queue()
+        self.sem = threading.BoundedSemaphore(value=2 * num_workers)
+
+        self._started = False
+        self._stop = threading.Event()
+        self._mp_stop = mp.Event()
+        self._read_thread = threading.Thread(
+            target=_populate_queue, args=(self.source, self.in_q, self._stop, self.sem)
+        )
+        self._method = method
+
+        self._map_threads: List[Union[threading.Thread, mp.Process]] = []
+
+    def iterator(self) -> Iterator[T]:
+        if not self._started:
+            for worker_id in range(self.num_workers):
+                args = (
+                    worker_id,
+                    self.in_q,
+                    self.out_q,
+                    self.in_order,
```

Comment:
Do I understand this correctly that all of the worker threads are getting the same `in_q` and `out_q`? That means that the parallel workers will compete with each other for work to do, and then compete with each other to submit processed work. (I don't mean this as a criticism, I'm just making sure I understand the logic correctly.)

Reply:
Yes that's how it's set up right now. I assume you mean in the sense of contention on the queue locks? When I last tested throughput for Imagenet, we were nowhere near hitting this limit (eg orders of magnitude lower QPS) but definitely open to changing this

Reply:
Yes, the "competition" will come through potential contention on the queue locks, but that will only make a difference if the work they're competing for is very, very small. I believe this is the correct pattern for when you have a relatively coarse granularity of work: a global queue, and workers take work when they're free to do so.